### PR TITLE
Add endpoints for CommentServlet and tests

### PR DESCRIPTION
**Pull request by lisasgoh**
Add doGet, doPost, doDelete and doPut endpoints as well as tests for these endpoints

### CODE REVIEW

File: src/main/java/com/google/step/servlets/CommentPostServlet.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/api/comments/")
+public class CommentPostServlet extends HttpServlet {
+
+  private CommentManager manager = new CommentManagerDatastore();
+
+  public CommentPostServlet(CommentManager commentManager) {
```
Comment (limli) [inline src/main/java/com/google/step/servlets/CommentPostServlet.java]: missing constructor without arguments. I believe the appengine won't work without it.

File: src/main/java/com/google/step/servlets/CommentPostServlet.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  /** Posts a comment for the deal with the given id parameter */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long dealId;
+    long userId;
+    String content;
+    try {
+      dealId = Long.parseLong(request.getParameter("dealId"));
+      userId = Long.parseLong(request.getParameter("userId"));
+      content = request.getParameter("content");
```
Comment (limli) [inline src/main/java/com/google/step/servlets/CommentPostServlet.java]: move the `content = request.getParameter("content");` outside of the try catch

File: src/main/java/com/google/step/servlets/CommentServlet.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    long id;
+    try {
+      id = Long.parseLong(request.getPathInfo().substring(1));
+    } catch (NumberFormatException e) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    } catch (StringIndexOutOfBoundsException e) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
```
Comment (jyx11011) [inline src/main/java/com/google/step/servlets/CommentServlet.java]: can catch both exceptions in just one catch statement

File: src/main/java/com/google/step/servlets/CommentServlet.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles individual comments. */
+@WebServlet("/api/comments/*")
+public class CommentServlet extends HttpServlet {
+
+  private CommentManager manager = new CommentManagerDatastore();
+
+  public CommentServlet(CommentManager commentManager) {
```
Comment (limli) [inline src/main/java/com/google/step/servlets/CommentServlet.java:18]: missing constructor without arguments

File: src/main/java/com/google/step/servlets/CommentServlet.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    manager = commentManager;
+  }
+
+  /** Deletes the comment with the given id parameter */
+  @Override
+  public void doDelete(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    long id;
+    try {
+      id = Long.parseLong(request.getPathInfo().substring(1));
```
Comment (limli) [inline src/main/java/com/google/step/servlets/CommentServlet.java:32]: might need to check if `request.getPathInfo()` is null, for when user goes to `/api/comments` directly without anything after /*

File: src/main/java/com/google/step/servlets/CommentServlet.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    manager.deleteComment(id);
+  }
+
+  /** Gets the comments for the deal with the given id parameter */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long dealId;
+    try {
+      dealId = Long.parseLong(request.getParameter("dealId"));
```
Comment (limli) [inline src/main/java/com/google/step/servlets/CommentServlet.java]: this servlet should be used for getting a single comment with commentId, because of its path `/api/comments/*`, not for getting all comments from a deal

File: src/main/java/com/google/step/servlets/JsonFormatter.java [hunk 7]
```diff
@@ -1,17 +1,48 @@
...
+import com.google.step.model.Comment;
 import com.google.step.model.Deal;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
 public class JsonFormatter {
+  public static String getCommentsJson(List<Comment> comments) {
+    if (comments.size() == 0) {
```
Comment (limli) [inline src/main/java/com/google/step/servlets/JsonFormatter.java]: I think the code below covers this case, so there is no need to explicitly cover it

**Review by limli — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/step/servlets/CommentServlet.java (hunk 3 above)
Comment (lisasgoh) [reply src/main/java/com/google/step/servlets/CommentServlet.java]: i removed the second catch instead cos i realised it wont be called

File: src/main/java/com/google/step/servlets/CommentServlet.java (hunk 5 above)
Comment (lisasgoh) [reply src/main/java/com/google/step/servlets/CommentServlet.java:32]: theres another servlet for that url (getpostservlet) i transferred the doget over

File: src/main/java/com/google/step/servlets/CommentServlet.java (hunk 6 above)
Comment (lisasgoh) [reply src/main/java/com/google/step/servlets/CommentServlet.java]: i dont think getting a single comment is necessary, so i just transferred the doget

File: src/main/java/com/google/step/servlets/CommentGetPostServlet.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  /** Posts a comment for the deal with the given id parameter */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    long dealId;
+    long userId;
+    String content;
+    try {
+      dealId = Long.parseLong(request.getParameter("dealId"));
+      userId = Long.parseLong(request.getParameter("userId"));
```
Comment (anuragshirolkar) [inline src/main/java/com/google/step/servlets/CommentGetPostServlet.java]: Is it possible to get the user id from the UserService? We should use the id of the logged in user rather than taking it from the request. It's possible that the user might manipulate the request to put someone else's user id.

File: src/main/java/com/google/step/servlets/CommentServlet.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+    content = request.getParameter("content");
+    Comment updatedComment = manager.updateComment(id, content);
+    if (updatedComment == null) {
+      response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+      return;
+    } else {
+      response.getWriter().println(JsonFormatter.getCommentJson(updatedComment));
```
Comment (anuragshirolkar) [inline src/main/java/com/google/step/servlets/CommentServlet.java:56]: Why is this needed?

File: src/test/java/com/google/step/servlets/CommentServletTest.java [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+  private static final long ID_B = 2;
+  private static final long USERID_B = 4;
+  private static final String CONTENT_B = "Hello world2";
+  private static final Comment COMMENT_B = new Comment(ID_B, DEALID, USERID_B, CONTENT_B);
+
+  private CommentManager commentManager;
+
+  private CommentServlet commentServlet;
+
+  private CommentGetPostServlet commentGetPostServlet;
```
Comment (anuragshirolkar) [inline src/test/java/com/google/step/servlets/CommentServletTest.java]: Could you split the CommentGetPostServlet's tests into a separate file? Sometimes people might forget which file has the tests, so it's better to create a separate test class corresponding to each servlet.

File: src/test/java/com/google/step/servlets/CommentServletTest.java [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+  private static final String UPDATE_CONTENT_A = "Update";
+  private static final Comment UPDATE_COMMENT_A =
+      new Comment(ID_A, DEALID, USERID_A, UPDATE_CONTENT_A);
+
+  private static final long ID_B = 2;
+  private static final long USERID_B = 4;
+  private static final String CONTENT_B = "Hello world2";
+  private static final Comment COMMENT_B = new Comment(ID_B, DEALID, USERID_B, CONTENT_B);
+
+  private CommentManager commentManager;
```
Comment (anuragshirolkar) [inline src/test/java/com/google/step/servlets/CommentServletTest.java]: Call this mockCommentManager.

**Review by anuragshirolkar — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/step/servlets/CommentGetPostServlet.java (hunk 8 above)
Comment (lisasgoh) [reply src/main/java/com/google/step/servlets/CommentGetPostServlet.java]: I set it as a TODO for now and hardcoded the value

File: src/main/java/com/google/step/servlets/CommentServlet.java (hunk 9 above)
Comment (lisasgoh) [reply src/main/java/com/google/step/servlets/CommentServlet.java:56]: i removed the redirect below and this part is mainly for testing

**Review by anuragshirolkar — APPROVED** [review]

**Review by limli — APPROVED** [review]
